docs(tutorial): add traffic light in planning simulation (#138)
autowarefoundation/autoware-documentation: docs/tutorials/ad-hoc-simulation/planning-simulation.md

2. In the 3D View pane, click and hold the left-mouse button, and then drag to set the direction for the goal pose. If done correctly, you will see a planned path from initial pose to goal pose.

![set-goal-pose](images/planning/lane-following/set-goal-pose.png)

#### Engage the ego vehicle

Now you can start the ego vehicle driving by clicking the `Engage` button in `AutowareStatePanel`. Alteratively, you can manually engage the vehicle by running the following command:

```bash
source ~/autoware/install/setup.bash
ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "engage: true" -1
```

![start-driving](images/planning/lane-following/engage-and-start-planning.png)

### Parking maneuvers

1. Set an initial pose and a goal pose, and engage the ego vehicle.

   ![after-set-goal-pose](images/planning/parking/after-set-goal-pose.png)

2. When the vehicle approaches the goal, it will change to the parking mode.
3. After that, the vehicle will start driving backwards and park at the destination parking spot.

   ![parking-maneuver](images/planning/parking/parking-maneuver.png)

## Advanced Simulations

> ### Dummy objects

Reviewer: Does it make it more clear to call this tutorial "Placing dummy objects"? Or is "Adding dummy objects" better?

```suggestion
### Placing dummy objects
```

Author: "Placing" would be good for the user

> To place a dummy car/pedestrian, click the `2D Dummy Car` and/or `2D Dummy Pedestrian` button in the toolbar, and then set its pose by dragging on the map. You can set its velocity in `Tool Properties -> 2D Dummy Car/Pedestrian` panel. In the image the velocity is set to 0.
> 
> ![set-dummy-car](images/planning/lane-following/place-dummy-car.png)
> 
> To delete any dummy objects placed in the view, click the `Delete All Objects` button in the toolbar.

Reviewer: - Broke this section up into individual steps and reworded it slightly.
- Previously, to delete dummy objects you had to click the `Delete All Objects` button and then click in the 3D view panel. Has that been changed?

```suggestion
a) Click the `2D Dummy Car` or `2D Dummy Pedestrian` button in the toolbar.
b) Set the pose of the dummy object by clicking and dragging on the map. 
c) Set the velocity of the object in `Tool Properties -> 2D Dummy Car/Pedestrian` panel. 

![set-dummy-car](images/planning/lane-following/place-dummy-car.png)

d) Delete any dummy objects placed in the view by clicking the `Delete All Objects` button in the toolbar.
```

### Traffic light simulation

#### Set traffic light

To simulate traffic light recognition, go to `Panels -> Add new panel`, select `TrafficLightPublishPanel`, and then press `OK`. Then in the panel you need to set the `ID` and color of the traffic light.

> You can check the ID of each traffic light on Rviz. If the IDs do not pop up, select `Map/Lanelet2VectorMap/Namespaces/traffic_light_id` topic, check the checkbox, and then **reload** this topic by clicking the checkbox of `Map` twice to re-visualize the IDs (see the image below). You can have a closer look at the IDs by zooming in the region or by chaning the View type.

Reviewer 2: typo?

```suggestion
You can check the ID of each traffic light on Rviz. If the IDs do not pop up, select `Map/Lanelet2VectorMap/Namespaces/traffic_light_id` topic, check the checkbox, and then **reload** this topic by clicking the checkbox of `Map` twice to re-visualize the IDs (see the image below). You can have a closer look at the IDs by zooming in the region or by changing the View type.
```

Reviewer 2: Resolved.

> To simulate traffic light recognition, go to `Panels -> Add new panel`, select `TrafficLightPublishPanel`, and then press `OK`. Then in the panel you need to set the `ID` and color of the traffic light.
> 
> You can check the ID of each traffic light on Rviz. If the IDs do not pop up, select `Map/Lanelet2VectorMap/Namespaces/traffic_light_id` topic, check the checkbox, and then **reload** this topic by clicking the checkbox of `Map` twice to re-visualize the IDs (see the image below). You can have a closer look at the IDs by zooming in the region or by changing the View type.
> 
> ![see-traffic-light-ID](images/planning/lane-following/see-traffic-light-ID.png)
> 
> And then you can select the ID of the desired traffic light from the `ID` _Combobox_. The image below shows a traffic light whose `ID` is 34836 and its color is set to `RED`.
> 
> ![set-traffic-light](images/planning/lane-following/set-traffic-light.png)
> 
> Finally, click the `PUBLISH` button to send the traffic light status to the simulator. Any planned path that goes past the selected traffic light will then change accordingly.
> 
> ![send-traffic-light-color](images/planning/lane-following/send-traffic-light-color.png)

Reviewer: - Reordered this section a lot, since I think it makes more sense to show the steps required to add the panel and set the ID and color before explaining how to find the IDs or troubleshooting the IDs not appearing.
- The `see-traffic-light-ID.png` seems to be used to show two different things (how to find the `traffic_light_id` topic for troubleshooting, and how to find the traffic light IDs on the map). I think it would be better to split the image so that we can clearly have one image to explain the troubleshooting and a different image to show how to find the IDs on the map. Any thoughts on this?
- Removed the description of the "set" screenshot since the screenshot is self-explanatory.
- For the Lane Driving tutorial, we have a screenshot showing how to add the AutowareStatePanel, so should we also add a similar screenshot showing how to add the TrafficLightPublishPanel?

```suggestion
a) Go to `Panels -> Add new panel`, select `TrafficLightPublishPanel`, and then press `OK`. 
b) In `TrafficLightPublishPanel`, set the `ID` and color of the traffic light.
c) Click the `SET` button.

![set-traffic-light](images/planning/lane-following/set-traffic-light.png)

d) Finally, click the `PUBLISH` button to send the traffic light status to the simulator. Any planned path that goes past the selected traffic light will then change accordingly.

![send-traffic-light-color](images/planning/lane-following/send-traffic-light-color.png)

By default, Rviz should display the ID of each traffic light on the map. You can have a closer look at the IDs by zooming in the region or by changing the View type.

In the event that the IDs are not displayed, try the following troubleshooting steps:
a) In the `Displays` panel, find the `traffic_light_id` topic by toggling the triangle icons next to `Map > Lanelet2VectorMap > Namespaces`. 
b) Check the `traffic_light_id` checkbox.
c) Reload the topic by clicking the `Map` checkbox twice.

![see-traffic-light-ID](images/planning/lane-following/see-traffic-light-ID.png)
```

#### Reset traffic light

> If you previously set a traffic light to `RED` or `AMBER`, you can reset the color of the traffic light to `GREEN` and let the ego vehicle make a turn at the intersection.

Reviewer: I note that the attached image has a red box around `SET` rather than `RESET`. Was this intentional?

```suggestion
If you set any traffic lights to `RED` or `AMBER` to confirm that the ego vehicle would correctly stop at an intersection, clicking the `RESET` button will reset the color of the traffic light to the default `GREEN` and let the ego vehicle continue along its planned path.
```

Author: Well 'RESET'ting actually deletes the traffic light information, so we need to 'SET'  its color to 'GREEN'. Yeah this explanation was ambiguous.

!!! warning

    Remember to set the initial pose of the car in the same direction as the lane.

    To confirm the direction of the lane, check the arrowheads displayed on the map.

#### 4. Set a goal pose for the ego vehicle

a) Click the `2D Goal Pose` button in the toolbar, or hit the `G` key.

b) In the 3D View pane, click and hold the left-mouse button, and then drag to set the direction for the goal pose. If done correctly, you will see a planned path from initial pose to goal pose.

![set-goal-pose](images/planning/lane-following/set-goal-pose.png)

#### 5. Engage the ego vehicle

Now you can start the ego vehicle driving by clicking the `Engage` button in `AutowareStatePanel`. Alteratively, you can manually engage the vehicle by running the following command:

```bash
source ~/autoware/install/setup.bash
ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "engage: true" -1
```

![start-driving](images/planning/lane-following/engage-and-start-planning.png)

### Parking scenario

1. Set an initial pose and a goal pose, and engage the ego vehicle.

   ![after-set-goal-pose](images/planning/parking/after-set-goal-pose.png)

> 2. When the vehicle approaches the goal, it will change to the parking mode.

Reviewer: ```suggestion
2. When the vehicle approaches the goal, it will switch from lane driving mode to parking mode.
```

3. After that, the vehicle will start driving backwards and park at the destination parking spot.

   ![parking-maneuver](images/planning/parking/parking-maneuver.png)

## Advanced Simulations

### Placing dummy objects

1. Click the `2D Dummy Car` or `2D Dummy Pedestrian` button in the toolbar.

2. Set the pose of the dummy object by clicking and dragging on the map.

3. Set the velocity of the object in `Tool Properties -> 2D Dummy Car/Pedestrian` panel.

> 

Reviewer: ```suggestion

!!! note

    Changes to the `velocity` parameter will only affect objects placed after the parameter is changed.
    
```

![set-dummy-car](images/planning/lane-following/place-dummy-car.png) 4. Delete any dummy objects placed in the view by clicking the `Delete All Objects` button in the toolbar.

### Traffic light recognition simulation

> 

Reviewer: My understanding is that traffic lights on the map are treated as green by default, so added some introductory text to explain that point. Please amend as necessary if my understanding is incorrect.

```suggestion
By default, traffic lights on the map are all treated as if they are set to green. As a result, when a path is created that passed through an intersection with a traffic light, the ego vehicle will drive through the intersection without stopping. 

The following steps explain how to set and reset traffic lights in order to test how the Planning component will respond.

```

Author: Yes that's right. This looks more clear.

#### Set traffic light

1. Go to `Panels -> Add new panel`, select `TrafficLightPublishPanel`, and then press `OK`.

2. In `TrafficLightPublishPanel`, set the `ID` and color of the traffic light.

3. Click the `SET` button.

> ![set-traffic-light](images/planning/lane-following/set-traffic-light.png) 4. Finally, click the `PUBLISH` button to send the traffic light status to the simulator. Any planned path that goes past the selected traffic light will then change accordingly.

Reviewer: ```suggestion
![set-traffic-light](images/planning/lane-following/set-traffic-light.png) 

4. Finally, click the `PUBLISH` button to send the traffic light status to the simulator. Any planned path that goes past the selected traffic light will then change accordingly.
```

Reviewer: I had a check, and it looks like the change happened here:
![image](https://user-images.githubusercontent.com/63835446/173083879-fecfcd10-1a96-4787-bfb6-7c5ddd3a51e0.png)

https://github.com/autowarefoundation/autoware-documentation/compare/1125f66f8cf18dad9907307f2e2ab85e678b09e1..d376f5d3739dbe600acd10b098c7325dde81ff25

![send-traffic-light-color](images/planning/lane-following/send-traffic-light-color.png)

By default, Rviz should display the ID of each traffic light on the map. You can have a closer look at the IDs by zooming in the region or by changing the View type.

In the event that the IDs are not displayed, try the following troubleshooting steps:

a) In the `Displays` panel, find the `traffic_light_id` topic by toggling the triangle icons next to `Map > Lanelet2VectorMap > Namespaces`.

b) Check the `traffic_light_id` checkbox.

c) Reload the topic by clicking the `Map` checkbox twice.

![see-traffic-light-ID](images/planning/lane-following/see-traffic-light-ID.png)

#### Update/Reset traffic light

You can update the color of the traffic light by selecting the next color (in the image it is `GREEN`) and clicking `SET` button. In the image the traffic light in front of the ego vehicle changed from `RED` to `GREEN` and the vehicle restarted.

![after-traffic-light-color-update](images/planning/lane-following/after-traffic-light-color-update.png)

> To delete a traffic light, use `RESET` button.

Reviewer: ```suggestion
To remove a traffic light from `TrafficLightPublishPanel`, click the `RESET` button. 
```

## Preparation

Download and unpack a sample map.

- Click [here](https://drive.google.com/file/d/1499_nsbUbIeturZaDj7jhUownh5fvXHd/view?usp=sharing) to download.
- Unpack it by running the following command.

```bash
unzip -d ~/Downloads/ ~/Downloads/sample-map-planning.zip
```

!!! Note

    Sample map: Copyright 2020 TIER IV, Inc.

## Basic Simulations

### Lane Driving

#### 1. Launch Autoware

> ```sh

Reviewer 2: Although this was originally wrong, could you replace `sh` with `bash` to be clear and consistent?

```suggestion
```bash
```

b) In the 3D View pane, click and hold the left-mouse button, and then drag to set the direction for the initial pose. An image representing the vehicle should now be displayed.

!!! warning

    Remember to set the initial pose of the car in the same direction as the lane.

    To confirm the direction of the lane, check the arrowheads displayed on the map.

#### 4. Set a goal pose for the ego vehicle

a) Click the `2D Goal Pose` button in the toolbar, or hit the `G` key.

b) In the 3D View pane, click and hold the left-mouse button, and then drag to set the direction for the goal pose. If done correctly, you will see a planned path from initial pose to goal pose.

![set-goal-pose](images/planning/lane-following/set-goal-pose.png)

#### 5. Engage the ego vehicle

Now you can start the ego vehicle driving by clicking the `Engage` button in `AutowareStatePanel`. Alteratively, you can manually engage the vehicle by running the following command:

```bash
source ~/autoware/install/setup.bash
ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "engage: true" -1
```

![start-driving](images/planning/lane-following/engage-and-start-planning.png)

> ### Parking maneuvers

Reviewer 2: [IMO] Using the words `Lane driving scenario` and `Parking scenario` might be better?

![image](https://user-images.githubusercontent.com/31987104/172384778-d027b2b3-c8d1-4b7c-9be9-2a52875a5d76.png)

Author: Oops thanks !

1. Set an initial pose and a goal pose, and engage the ego vehicle.

   ![after-set-goal-pose](images/planning/parking/after-set-goal-pose.png)

2. When the vehicle approaches the goal, it will change to the parking mode.
3. After that, the vehicle will start driving backwards and park at the destination parking spot.

   ![parking-maneuver](images/planning/parking/parking-maneuver.png)

## Advanced Simulations

### Placing Dummy objects

> a) Click the `2D Dummy Car` or `2D Dummy Pedestrian` button in the toolbar.

Reviewer 2: Why did you use `a, b, c, d` here instead of `1, 2, 3, 4`? :thinking: 
If there is no specific reason, it's better to make the style consistent.

#### 5. Engage the ego vehicle

Now you can start the ego vehicle driving by clicking the `Engage` button in `AutowareStatePanel`. Alteratively, you can manually engage the vehicle by running the following command:

```bash
source ~/autoware/install/setup.bash
ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "engage: true" -1
```

![start-driving](images/planning/lane-following/engage-and-start-planning.png)

### Parking scenario

1. Set an initial pose and a goal pose, and engage the ego vehicle.

   ![after-set-goal-pose](images/planning/parking/after-set-goal-pose.png)

2. When the vehicle approaches the goal, it will switch from lane driving mode to parking mode.
3. After that, the vehicle will reverse into the destination parking spot.

   ![parking-maneuver](images/planning/parking/parking-maneuver.png)

## Advanced Simulations

### Placing dummy objects

1. Click the `2D Dummy Car` or `2D Dummy Pedestrian` button in the toolbar.

2. Set the pose of the dummy object by clicking and dragging on the map.

3. Set the velocity of the object in `Tool Properties -> 2D Dummy Car/Pedestrian` panel.

> !!! note
> Changes to the `velocity` parameter will only affect objects placed after the parameter is changed.

Reviewer: Fixing the note admonition.

```suggestion
!!! note

    Changes to the `velocity` parameter will only affect objects placed after the parameter is changed.
```

Author: Thank you!
